### PR TITLE
Reuse mimirpb.WriteRequest marshaled bytes

### PR DESCRIPTION
**Pull request by colega**
#### What this PR does

I realized that most of the allocated bytes in the distributor come from `mimirpb.WriteRequest.Marshal()`. It's true that it's not a lot of allocations, a few per each push, but a lot of bytes are being moved there, so I decided to try to pool those.

The side effects of this (on the following graphs, I deployed this change around 12:05 UTC):

CPU dropped around 15%, from avg 5.1 to 4.3 cores in our dev cell.

image

Heap inuse has grown ~45%:

image

And workingset doubled:

image

Yet, all the memory stays within the limits.

It seems that there's a benefit in CPU and we could protect from OOMing by using GOMEMLIMIT there, with an overall improvement.

Leaving this draft because I want to see people's opinion (@bboreham, @pracucci?) first, before talking about actual code changes (although feel free to leave comments on the code too).

#### Which issue(s) this PR fixes or relates to

None, was just passing by.

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by bboreham** [comment]
> not a lot of allocations, a few per each push, but a lot of bytes 

Generally the rate of bytes allocated is what drives the cost of garbage-collection.
Each allocation is very quick, just flipping a bit, but each garbage-collect requires following every pointer in every live object.

The cost of one GC is proportional to the number of live pointers, and the cost of GC overall is that times the rate of GC, which is proportional to the rate of bytes allocated (for a given heap size).

File: pkg/mimirpb/write_request.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+package mimirpb
+
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+var reusableWriteRequestsPool = sync.Pool{
```
Comment (bboreham) [inline pkg/mimirpb/write_request.go:11]: Consider 	github.com/grafana/mimir/pkg/util/pool.BucketedBytes.

**Comment by pracucci** [comment]
I think we should this optimization in terms on impact on TCO. I reach you out on Slack so we can discuss it further.

File: pkg/mimirpb/write_request.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+func NewReusableWriteRequest(request WriteRequest) *ReusableWriteRequest {
+	wr := reusableWriteRequestsPool.Get().(*ReusableWriteRequest)
+	wr.WriteRequest = request
+	return wr
+}
+
+type ReusableWriteRequest struct {
+	WriteRequest
+
+	bytes atomic.Pointer[[]byte]
```
[the remark this answers is not in the document]
Comment (colega) [reply pkg/mimirpb/write_request.go]: Removed the atomic.

**Comment by colega** [comment]
I've been running this the entire weekend, from on two deployments on our dev env, 3 replicas each.

The deployment has been running from March 31th ~10:00 UTC to April 3rd ~7.30 UTC

The results are strange and interesting. 

### CPU

[CPU usage](https://admin-dev-us-central-0.grafana.net/grafana/explore?left=%7B%22datasource%22:%22P09658BD21492D183%22,%22queries%22:%5B%7B%22expr%22:%22sum%20by%28deployment%29%20%28%5Cn%20%20%20%20label_replace%28%5Cn%20%20%20%20%20%20%20%20rate%28container_cpu_usage_seconds_total%7Bnamespace%3D~%5C%22cortex-dev-01%5C%22,pod%3D~%5C%22distributor.%2B%5C%22%7D%5B$__rate_interval%5D%29,%5Cn%20%20%20%20%20%20%20%20%5C%22deployment%5C%22,%20%5C%22$1%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22pod%5C%22,%20%5C%22%28.%2B%29-%5C%5C%5C%5Cw%2B-%5C%5C%5C%5Cw%2B%5C%22%5Cn%20%20%20%20%29%20%20%20%20%20%20%20%5Cn%29%22,%22format%22:%22time_series%22,%22intervalFactor%22:2,%22legendLink%22:null,%22step%22:10,%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P09658BD21492D183%22%7D,%22interval%22:%22%22,%22editorMode%22:%22code%22,%22range%22:true,%22instant%22:true,%22hide%22:false%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P09658BD21492D183%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%28deployment%29%20%28%5Cn%20%20%20%20label_replace%28%5Cn%20%20%20%20%20%20%20%20rate%28container_cpu_usage_seconds_total%7Bnamespace%3D~%5C%22cortex-dev-01%5C%22,pod%3D~%5C%22distributor.%2B%5C%22%7D%5B10m%5D%29,%5Cn%20%20%20%20%20%20%20%20%5C%22deployment%5C%22,%20%5C%22$1%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22pod%5C%22,%20%5C%22%28.%2B%29-%5C%5C%5C%5Cw%2B-%5C%5C%5C%5Cw%2B%5C%22%5Cn%20%20%20%20%29%20%20%20%20%20%20%20%5Cn%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D%5D,%22range%22:%7B%22from%22:%221680256800000%22,%22to%22:%221680501600000%22%7D%7D&orgId=1) is ~10% less _most of the time_:

image

However, during some intervals of time, it's actually 10% higher:

image

☝️  This specific interval correponds to an interval where the rate of the requests grew on both deployments.

Some large spans of time look really good, like these 12h:

image

Others however are a tight tie:

image

I can't correlate these to any change in the write requests rate:

image

### Memory

Memory results aren't that surprising: we're retaining the memory after all, so I expect it to be higher but less spiky.

[Workingset is slightly higher](https://admin-dev-us-central-0.grafana.net/grafana/explore?left=%7B%22datasource%22:%22P09658BD21492D183%22,%22queries%22:%5B%7B%22expr%22:%22sum%20by%28deployment%29%20%28%5Cn%20%20%20%20label_replace%28%5Cn%20%20%20%20%20%20%20%20max%20by%28pod%29%20%28container_memory_working_set_bytes%7Bnamespace%3D~%5C%22cortex-dev-01%5C%22,%20pod%3D~%5C%22distributor.%2B%5C%22%7D%29,%5Cn%20%20%20%20%20%20%20%20%5C%22deployment%5C%22,%20%5C%22$1%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22pod%5C%22,%20%5C%22%28.%2B%29-%5C%5C%5C%5Cw%2B-%5C%5C%5C%5Cw%2B%5C%22%5Cn%20%20%20%20%29%20%20%20%20%20%20%20%5Cn%29%20%2F%201e6%22,%22format%22:%22time_series%22,%22intervalFactor%22:2,%22legendLink%22:null,%22step%22:10,%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P09658BD21492D183%22%7D,%22interval%22:%22%22,%22editorMode%22:%22code%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221680256800000%22,%22to%22:%221680501600000%22%7D%7D&orgId=1)

image

[Heap inuse](https://admin-dev-us-central-0.grafana.net/grafana/explore?left=%7B%22datasource%22:%22P09658BD21492D183%22,%22queries%22:%5B%7B%22expr%22:%22sum%20by%28deployment%29%20%28%5Cn%20%20%20%20label_replace%28go_memstats_heap_inuse_bytes%7Bnamespace%3D~%5C%22cortex-dev-01%5C%22,pod%3D~%5C%22distributor.%2B%5C%22%7D,%5Cn%20%20%20%20%20%20%20%20%5C%22deployment%5C%22,%20%5C%22$1%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22pod%5C%22,%20%5C%22%28.%2B%29-%5C%5C%5C%5Cw%2B-%5C%5C%5C%5Cw%2B%5C%22%5Cn%20%20%20%20%29%20%20%20%20%20%20%20%5Cn%29%20%2F%201e6%22,%22format%22:%22time_series%22,%22intervalFactor%22:2,%22legendLink%22:null,%22step%22:10,%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P09658BD21492D183%22%7D,%22interval%22:%22%22,%22editorMode%22:%22code%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D,%7B%22expr%22:%22avg_over_time%28%28sum%20by%28deployment%29%20%28%5Cn%20%20%20%20label_replace%28go_memstats_heap_inuse_bytes%7Bnamespace%3D~%5C%22cortex-dev-01%5C%22,pod%3D~%5C%22distributor.%2B%5C%22%7D,%5Cn%20%20%20%20%20%20%20%20%5C%22deployment%5C%22,%20%5C%22$1%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22pod%5C%22,%20%5C%22%28.%2B%29-%5C%5C%5C%5Cw%2B-%5C%5C%5C%5Cw%2B%5C%22%5Cn%20%20%20%20%29%20%20%20%20%20%20%20%5Cn%29%20%2F%201e6%29%5B1h:%5D%29%22,%22format%22:%22time_series%22,%22intervalFactor%22:2,%22legendLink%22:null,%22step%22:10,%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P09658BD21492D183%22%7D,%22interval%22:%22%22,%22editorMode%22:%22code%22,%22range%22:true,%22instant%22:true,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221680256800000%22,%22to%22:%221680501600000%22%7D%7D&orgId=1) is higher on some 1h intervals:

image

As expected, when zoomed into smaller time intevals with higher resolution, it's less spiky:

image

**Comment by lamida** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

**Comment by colega** [comment]
The results are unconclusive. Closing until we're interested in this again.

**Pull request closed without merging**